types/lib/elements/dom-repeat.d.ts

```diff
@@ -0,0 +1,268 @@
+   * `this.$.domRepeat.modelForElement(event.target).set('item.checked', true);`.
+   * This may be useful for manipulating instance data of event targets obtained
+   * by event handlers on parents of the `dom-repeat` (event delegation).
+   *
+   * A view-specific filter/sort may be applied to each `dom-repeat` by supplying a
+   * `filter` and/or `sort` property.  This may be a string that names a function on
+   * the host, or a function may be assigned to the property directly.  The functions
+   * should implemented following the standard `Array` filter/sort API.
+   *
+   * In order to re-run the filter or sort functions based on changes to sub-fields
+   * of `items`, the `observe` property may be set as a space-separated list of
+   * `item` sub-fields that should cause a re-filter/sort when modified.  If
+   * the filter or sort function depends on properties not contained in `items`,
+   * the user should observe changes to those properties and call `render` to update
+   * the view based on the dependency change.
+   *
+   * For example, for an `dom-repeat` with a filter of the following:
+   *
+   * ```js
+   * isEngineer: function(item) {
+   *     return item.type == 'engineer' || item.manager.type == 'engineer';
+   * }
+   * ```
+   *
+   * Then the `observe` property should be configured as follows:
+   *
+   * ```html
+   * <template is="dom-repeat" items="{{employees}}"
+   *           filter="isEngineer" observe="type manager.type">
+   * ```
+   */
+  class DomRepeat extends
+    Polymer.OptionalMutableData(
+    Polymer.Element) {
+
+    /**
+     * An array containing items determining how many instances of the template
+     * to stamp and that that each template instance should bind to.
+     */
+    items: any[]|null;
+
+    /**
+     * The name of the variable to add to the binding scope for the array
+     * element associated with a given template instance.
+     */
+    as: string;
+
+    /**
+     * The name of the variable to add to the binding scope with the index
+     * of the instance in the sorted and filtered list of rendered items.
+     * Note, for the index in the `this.items` array, use the value of the
+     * `itemsIndexAs` property.
+     */
+    indexAs: string;
+
+    /**
+     * The name of the variable to add to the binding scope with the index
+     * of the instance in the `this.items` array. Note, for the index of
+     * this instance in the sorted and filtered list of rendered items,
+     * use the value of the `indexAs` property.
+     */
+    itemsIndexAs: string;
+
+    /**
+     * A function that should determine the sort order of the items.  This
+     * property should either be provided as a string, indicating a method
+     * name on the element's host, or else be an actual function.  The
+     * function should match the sort function passed to `Array.sort`.
+     * Using a sort function has no effect on the underlying `items` array.
+     */
+    sort: Function|null;
```

Comment:
TODO Update how Analyzer emits typings for Polymer properties. Talked to @rictic, we're thinking:

- If there is a `@type` annotation, emit it (already the case).
- If there is a default `value`, emit such that the type is non-nullable and non-undefined.
- Otherwise, assume the type could be null or undefined.

Reply:
Doing this right would probably be a change to Analyzer. Draft PR out at https://github.com/Polymer/polymer-analyzer/pull/791. Needs a bit more discussion. Proposal is that all Polymer properties should be emitted as possibly null or undefined, unless they have a `@type` annotation to constrain it further.

Reply:
Done via https://github.com/Polymer/polymer-analyzer/pull/791. Properties are null always `|null|undefined` unless a `@type` annotation says otherwise.